### PR TITLE
REGRESSION(262173@main): media/media-source/media-source-fastseek.html is a flaky text failure

### PR DESCRIPTION
**Pull request by jyavenard**
#### ab74fe9cba783e46bf6dc8d39e98215d383805c3
<pre>
Need a short description (OOPS!).
Need the bug URL (OOPS!).
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::completeSeek):
</pre>

### CODE REVIEW

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 1]
```diff
@@ -177,52 +176,47 @@ MediaTime MediaSource::duration() const
 
 MediaTime MediaSource::currentTime() const
 {
-    return m_mediaElement ? m_mediaElement->currentMediaTime() : MediaTime::zeroTime();
+    return m_pendingSeekTarget ? m_pendingSeekTarget->time : m_mediaElement ? m_mediaElement->currentMediaTime() : MediaTime::zeroTime();
```
Comment (eric-carlson) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp]: Nit: I think nested ternary operators are hard to read, especially when they get so wide you that you have to scroll horizontally to see the whole thing.

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 2]
```diff
@@ -234,24 +228,52 @@ void MediaSource::completeSeek()
...
     // with the closest random access point before the new playback position.
-    MediaTime pendingSeekTime = m_pendingSeekTime;
-    m_pendingSeekTime = MediaTime::invalidTime();
-    m_private->setIsSeeking(false);
-    for (auto& sourceBuffer : *m_activeSourceBuffers)
-        sourceBuffer->seekToTime(pendingSeekTime);
+    auto seekTarget = *m_pendingSeekTarget;
+    m_pendingSeekTarget.reset();
+
+    struct CallbackAggregator final : public RefCounted<CallbackAggregator> {
```
Comment (eric-carlson) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp]: We already have <wtf/CallbackAggregator.h> so this should probably have a more specific name.

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h [hunk 3]
```diff
@@ -85,8 +85,8 @@ ALLOW_NEW_API_WITHOUT_GUARDS_END
     void removeVideoTrack(VideoTrackPrivate&);
     void removeTextTrack(InbandTextTrackPrivate&);
 
-    MediaPlayer::NetworkState networkState() const override;
-    MediaPlayer::ReadyState readyState() const override;
+    MediaPlayer::NetworkState networkState() const final;
+    MediaPlayer::ReadyState readyState() const final;
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h]: These unrelated `override` to `final` changes are a mistake:
- they make it harder to pick out and review the changes necessary for this fix
- they will make future `blame` analysis more complicated
- I believe it is contrary to standard WebKit practice

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h [hunk 4]
```diff
@@ -72,16 +72,16 @@ class MediaPlayerPrivateMediaStreamAVFObjC final
     static void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types);
     static MediaPlayer::SupportsType supportsType(const MediaEngineSupportParameters&);
 
-    MediaPlayer::NetworkState networkState() const override;
+    MediaPlayer::NetworkState networkState() const final;
     void setNetworkState(MediaPlayer::NetworkState);
-    MediaPlayer::ReadyState readyState() const override;
+    MediaPlayer::ReadyState readyState() const final;
```
Comment (eric-carlson) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h]: Ditto.

File: Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h [hunk 5]
```diff
@@ -50,47 +50,45 @@ class MockMediaPlayerMediaSource : public MediaPlayerPrivateInterface, public Ca
...
-    void pause() override;
-    FloatSize naturalSize() const override;
-    bool hasVideo() const override;
-    bool hasAudio() const override;
+    void cancelLoad() final;
+    void play() final;
+    void pause() final;
+    FloatSize naturalSize() const final;
+    bool hasVideo() const final;
+    bool hasAudio() const final;
```
Comment (eric-carlson) [inline Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h]: Ditto

File: Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h [hunk 6]
```diff
@@ -79,16 +78,14 @@ class MockMediaSourcePrivate final
...
-    void waitForSeekCompleted() override;
-    void seekCompleted() override;
+    // MediaSourcePrivate finals
+    AddStatus addSourceBuffer(const ContentType&, bool webMParserEnabled, RefPtr<SourceBufferPrivate>&) final;
+    void durationChanged(const MediaTime&) final;
+    void markEndOfStream(EndOfStreamStatus) final;
+    bool isEnded() const final { return m_isEnded; }
+    void unmarkEndOfStream() final;
+    MediaPlayer::ReadyState readyState() const final;
+    void setReadyState(MediaPlayer::ReadyState) final;
```
Comment (eric-carlson) [inline Source/WebCore/platform/mock/mediasource/MockMediaSourcePrivate.h]: Ditto

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 2 above)
Comment (jyavenard) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp]: I've followed the existing pattern, there are multiple CallbackAggregator object in the code, they all replicate the WTF one with some changes.
There's one in WebSiteDataStore, 3 in the NetworkProcess code, and on in CDMInstanceFairPlayStreamingAVFObjC, and a few more with a slight change in the name

Comment (jyavenard) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp]: I renamed it to `SeeksCallbackAggregator`

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h (hunk 3 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h]: done

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 1 above)
Comment (jyavenard) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp]: broke it a bit

File: Source/WebCore/platform/graphics/MediaPlayer.cpp [hunk 7]
```diff
@@ -2013,6 +2013,17 @@ String MediaPlayer::lastErrorMessage() const
     return m_lastErrorMessage;
 }
 
+String SeekTarget::toString() const
+{
+    StringBuilder builder;
+    builder.append("[");
+    builder.append(WTF::LogArgument<MediaTime>::toString(time));
+    builder.append(WTF::LogArgument<MediaTime>::toString(negativeThreshold));
+    builder.append(WTF::LogArgument<MediaTime>::toString(positiveThreshold));
```
Comment (squelart) [inline Source/WebCore/platform/graphics/MediaPlayer.cpp:2022]: Shouldn't you also append some separators between these values?

Comment (jyavenard) [reply Source/WebCore/platform/graphics/MediaPlayer.cpp:2022]: it shows like this `[{"value":2701.603090079002}{"value":2}{"value":5}]` which I thought was good enough as MediaTime already adds `{ }`

Comment (squelart) [reply Source/WebCore/platform/graphics/MediaPlayer.cpp:2022]: Alright then, good enough for logging/debugging by people in the know, I suppose. 👍

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 8]
```diff
@@ -185,73 +186,98 @@ const PlatformTimeRanges& MediaSource::buffered() const
...
-    MediaTime pendingSeekTime = m_pendingSeekTime;
-    m_pendingSeekTime = MediaTime::invalidTime();
-    m_private->setIsSeeking(false);
-    for (auto& sourceBuffer : *m_activeSourceBuffers)
-        sourceBuffer->seekToTime(pendingSeekTime);
+    auto seekTarget = *m_pendingSeekTarget;
+    m_pendingSeekTarget.reset();
+
+    struct SeeksCallbackAggregator final : public RefCounted<SeeksCallbackAggregator> {
+        explicit SeeksCallbackAggregator(MediaTime target, MediaSource& source, CompletionHandler<void(const MediaTime&)>&& completionHandler)
```
Comment (youennf) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp]: No need for explicit

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 9]
```diff
@@ -185,73 +186,98 @@ const PlatformTimeRanges& MediaSource::buffered() const
...
+            : time(target)
+            , mediaSource(source)
+            , completionHandler(WTFMove(completionHandler))
+        {
+        }
+
+        ~SeeksCallbackAggregator()
+        {
+            ASSERT(completionHandler);
```
Comment (youennf) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp]: Does not bring much here, maybe add it in constructor instead if useful?

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 10]
```diff
@@ -185,73 +186,98 @@ const PlatformTimeRanges& MediaSource::buffered() const
     return m_buffered;
 }
 
-void MediaSource::seekToTime(const MediaTime& time)
+void MediaSource::seekToTarget(const SeekTarget& target, CompletionHandler<void(const MediaTime&)>&& completionHandler)
 {
     if (isClosed())
         return;
```
Comment (youennf) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp:192]: Should we `seekCompletedHandler(MediaTime::invalidTime())` if closed?

File: Source/WebCore/Modules/mediasource/MediaSource.cpp [hunk 11]
```diff
@@ -978,6 +1004,9 @@ void MediaSource::detachFromElement(HTMLMediaElement& element)
 
     m_private = nullptr;
     m_mediaElement = nullptr;
+
+    if (m_seekCompletedHandler)
+        m_seekCompletedHandler(MediaTime::invalidTime());
```
Comment (youennf) [inline Source/WebCore/Modules/mediasource/MediaSource.cpp:1011]: Is it the only case where we need to call m_seekCompletedHandler?
What about when it gets stopped for instance?

File: Source/WebCore/html/HTMLMediaElement.h [hunk 12]
```diff
@@ -1081,19 +1081,15 @@ class HTMLMediaElement
...
         {
         }
         MediaTime now;
-        MediaTime targetTime;
-        MediaTime negativeTolerance;
-        MediaTime positiveTolerance;
+        SeekTarget target;
     };
-    std::unique_ptr<PendingSeek> m_pendingSeek;
+    std::optional<PendingSeek> m_pendingSeek;
```
Comment (youennf) [inline Source/WebCore/html/HTMLMediaElement.h]: We have done some work to move from optional to unique_ptr to reduce some object's memory footprint.
Do we need this change?

File: Source/WebCore/platform/graphics/MediaPlayerPrivate.h [hunk 13]
```diff
@@ -20,7 +20,7 @@
  * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
  * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Comment (youennf) [inline Source/WebCore/platform/graphics/MediaPlayerPrivate.h:23]: All these file changes are unrelated.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp [hunk 14]
```diff
@@ -256,6 +256,27 @@ void SourceBufferPrivate::reenqueSamples(const AtomString& trackID)
     reenqueueMediaForTime(*trackBuffer, trackID, currentMediaTime());
 }
 
+void SourceBufferPrivate::seekToTarget(const SeekTarget& target, CompletionHandler<void(const MediaTime&)>&& completionHandler)
+{
+    if (!isAttached())
+        return;
```
Comment (youennf) [inline Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:263]: Call completionHandler probably.

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm [hunk 15]
```diff
@@ -529,81 +521,66 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
...
-    if (!pendingSeek->negativeThreshold && !pendingSeek->positiveThreshold)
-        m_lastSeekTime = pendingSeek->targetTime;
-    else
-        m_lastSeekTime = m_mediaSourcePrivate->fastSeekTimeForMediaTime(pendingSeek->targetTime, pendingSeek->positiveThreshold, pendingSeek->negativeThreshold);
-
-    if (m_lastSeekTime.hasDoubleValue())
-        m_lastSeekTime = MediaTime::createWithDouble(m_lastSeekTime.toDouble(), MediaTime::DefaultTimeScale);
-
-    m_seekCompleted = WaitingToSeek;
+    auto pendingSeek = WTFMove(*m_pendingSeek);
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm]: Can we std::exchange?

File: Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm [hunk 16]
```diff
@@ -529,81 +521,66 @@ void getSupportedTypes(HashSet<String, ASCIICaseInsensitiveHash>& types) const f
...
     ALWAYS_LOG(LOGIDENTIFIER);
-    m_seekCompleted = SeekCompleted;
+    m_seekState = SeekCompleted;
     if (shouldBePlaying())
         [m_synchronizer setRate:m_rate];
     if (!m_synchronizerSeeking) {
         if (auto player = m_player.get())
             player->timeChanged();
-    }
+    } else
```
Comment (youennf) [inline Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm]: Would read better with
```
if () {
    ...
    return;
}
ALWAYS_LOG(...)

File: Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp [hunk 17]
```diff
@@ -83,12 +83,14 @@ const PlatformTimeRanges& RemoteMediaSourceProxy::buffered() const
...
+void RemoteMediaSourceProxy::seekToTarget(const WebCore::SeekTarget& target, CompletionHandler<void(const MediaTime&)>&& completionHandler)
 {
-    if (!m_connectionToWebProcess)
+    if (!m_connectionToWebProcess) {
+        completionHandler(MediaTime::invalidTime());
         return;
+    }
 
-    m_connectionToWebProcess->connection().send(Messages::MediaSourcePrivateRemote::SeekToTime(time), m_identifier);
+    m_connectionToWebProcess->connection().sendWithAsyncReply(Messages::MediaSourcePrivateRemote::SeekToTarget(target), WTFMove(completionHandler), m_identifier);
```
Comment (youennf) [inline Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:93]: Is MediaTime:: invalidTime() used if GPUProcess crashes in the middle of this async IPC?

File: Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp [hunk 18]
```diff
@@ -355,13 +355,15 @@ void SourceBufferPrivateRemote::setAppendWindowEnd(const MediaTime& appendWindow
...
 {
     auto gpuProcessConnection = m_gpuProcessConnection.get();
-    if (!isGPURunning())
+    if (!isGPURunning()) {
+        completionHandler(MediaTime::invalidTime());
         return;
+    }
 
-    gpuProcessConnection->connection().send(Messages::RemoteSourceBufferProxy::SeekToTime(mediaTime), m_remoteSourceBufferIdentifier);
+    gpuProcessConnection->connection().sendWithAsyncReply(Messages::RemoteSourceBufferProxy::SeekToTarget(target), WTFMove(completionHandler), m_remoteSourceBufferIdentifier);
```
Comment (youennf) [inline Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:366]: Ditto, invalidTime is probably what we want if GPUProcess crashes within the async call.

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 10 above)
Comment (jyavenard) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp:192]: This is done when the MediaSource readyState changes to closed.

File: Source/WebCore/html/HTMLMediaElement.h (hunk 12 above)
Comment (jyavenard) [reply Source/WebCore/html/HTMLMediaElement.h]: I wasn't aware of this. I would have assumed here that allocation within the HTMMediaElement class is better than a single heap allocation with all the required overhead.

File: Source/WebCore/platform/graphics/SourceBufferPrivate.cpp (hunk 14 above)
Comment (jyavenard) [reply Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:263]: Ah indeed

File: Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp (hunk 17 above)
Comment (jyavenard) [reply Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:93]: I don't know how IPC deals with process crashes with completion handler. This seems to be a more overall problem.

File: Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp (hunk 18 above)
Comment (jyavenard) [reply Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:366]: I've created a `AsyncReplyError<WTF::MediaTime>` that sets the default return value to `MediaTime::invalidTime()`

thanks for the pointers

**Review by youennf — APPROVED** [review]
r=me once bots are happy

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 10 above)
Comment (jyavenard) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp:192]: As discussed, added the `AsyncReplyError<WTF::MediaTime>` template to handle this

File: Source/WebCore/Modules/mediasource/MediaSource.cpp (hunk 11 above)
Comment (jyavenard) [reply Source/WebCore/Modules/mediasource/MediaSource.cpp:1011]: done

File: Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp (hunk 17 above)
Comment (jyavenard) [reply Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:93]: Created a `AsyncReplyError<WTF::MediaTime>` template to define the default value when a connection failure occurs
